File: groking_algorithm/quick_sort.py
def partition(list, lo, hi):
  i = lo + 1
  pivot = list[lo]
  for j in range(lo + 1, hi + 1):
    if list[j] < pivot:
      list[i], list[j] = list[j], list[i]
      i = i + 1

  list[lo], list[i-1] = list[i-1], list[lo]
  return i-1

def quick_sort(list, lo, hi):
  if (lo < hi):
    p = partition(list, lo, hi)
    quick_sort(list, lo, p-1)
    quick_sort(list, p+1, hi)

arr = [1,12,5,26,7,14,3,7,2]
arr = [10, 7, 8, 9, 1, 5]

print('original: %s' % arr)
quick_sort(arr, 0, len(arr)-1)
print(arr)

File: groking_algorithm/quick_sort.js
function partition(list, lo, hi)
{
  let i = lo + 1;
  let pivot = list[lo].key;
  for (let j = lo + 1; j < hi + 1; ++j) {
    if (list[j].key < pivot) {
      let clone = list[i].key;
      list[i].key = list[j].key;
      list[j].key = clone;
      i++;
    }
  }

  let clone = list[lo].key;
  list[lo].key = list[i-1].key;
  list[i-1].key = clone;
  return i - 1;
}

function quicksort(list, lo, hi)
{
  if (lo < hi) {
    let p = partition(list, lo, hi);
    quicksort(list, lo, p-1);
    quicksort(list, p+1, hi);
  }
}

let arr = [1,12,5,26,7,14,3,7,2];
let list = [
  {key: 'Devin'},
  {key: 'Jackson'},
  {key: 'James'},
  {key: 'Joel'},
  {key: 'John'},
  {key: 'Jillian'},
  {key: 'Jimmy'},
  {key: 'Julie'},
]

console.log(`----original:`);
for (let e of list) {
  console.log(e);
}
quicksort(list, 0, list.length-1);

console.log(`----after:`);
for (let e of list) {
  console.log(e);
}
